File: backend/rorapp/helpers/text.py
from typing import List


def to_sentence_case(text: str) -> str:
    if not text:
        return text
    return text[0].upper() + text[1:]


def pluralize(count: int, singular: str, plural: str | None = None) -> str:
    """Return 'count word' with correct singular/plural form."""
    word = (
        singular if count == 1 else (plural if plural is not None else singular + "s")
    )
    return f"{count} {word}"


def possessive(name: str) -> str:
    """Return the possessive form of a name"""
    return f"{name}'" if name.endswith("s") else f"{name}'s"


def to_family_adjective(family_name: str) -> str:
    """Return the adjectival form of a Roman family name."""
    if family_name.endswith("us"):
        return family_name[:-2] + "an"
    return family_name


def format_list(items: List[str]) -> str:
    """
    Format a list of strings with commas and 'and' before the last item.
    """

    if not items:
        return ""
    if len(items) == 1:
        return items[0]
    return " and ".join(", ".join(items).rsplit(", ", 1))

File: frontend/helpers/text.ts
/**
 * Converts a string to its possessive form by adding 's or ' depending on the last letter.
 */
export function toPossessive(str: string): string {
  if (!str || str.length === 0) {
    return str
  }
  if (str.toLowerCase().endsWith("s")) {
    return str + "'"
  }
  return str + "'s"
}

/**
 * Format a list of strings with commas and 'and' before the last item.
 */
export function formatList(items: string[]): string {
  if (items.length === 0) return ""
  if (items.length === 1) return items[0]
  if (items.length === 2) return `${items[0]} and ${items[1]}`
  return `${items.slice(0, -1).join(", ")}, and ${items[items.length - 1]}`
}

/**
 * Capitalize the first letter.
 */
export function toSentenceCase(str: string): string {
  if (!str || str.length === 0) {
    return str
  }
  return str.charAt(0).toUpperCase() + str.slice(1)
}

/**
 * Converts a Roman family name to its adjectival form.
 */
export function toFamilyAdjective(familyName: string): string {
  if (familyName.endsWith("us")) {
    return familyName.slice(0, -2) + "an"
  }
  return familyName
}
